ios/browser/api/bookmarks/exporter/bookmark_html_writer.cc

```diff
@@ -0,0 +1,636 @@
+      NotifyOnFinish(BookmarksExportObserver::Result::kCouldNotWriteNodes);
+      return;
+    }
+
+    DecrementIndent();
+
+    Write(kFolderChildrenEnd);
+    Write(kNewline);
+    // File close is forced so that unit test could read it.
+    file_.reset();
+
+    NotifyOnFinish(BookmarksExportObserver::Result::kSuccess);
+  }
+
+ private:
+  friend class base::RefCountedThreadSafe<Writer>;
+
+  // Types of text being written out. The type dictates how the text is
+  // escaped.
+  enum TextType {
+    // The text is the value of an html attribute, eg foo in
+    // <a href="foo">.
+    ATTRIBUTE_VALUE,
+
+    // Actual content, eg foo in <h1>foo</h2>.
+    CONTENT
+  };
+
+  ~Writer() {}
+
+  // Opens the file, returning true on success.
+  bool OpenFile() {
+    int flags = base::File::FLAG_CREATE_ALWAYS | base::File::FLAG_WRITE;
+    file_.reset(new base::File(path_, flags));
+    if (!file_->IsValid()) {
+      PLOG(ERROR) << "Could not create " << path_;
+      return false;
+    }
+    return true;
+  }
+
+  // Increments the indent.
+  void IncrementIndent() {
+    indent_.resize(indent_.size() + kIndentSize, ' ');
+  }
+
+  // Decrements the indent.
+  void DecrementIndent() {
+    DCHECK(!indent_.empty());
+    indent_.resize(indent_.size() - kIndentSize, ' ');
+  }
+
+  // Called at the end of the export process.
+  void NotifyOnFinish(BookmarksExportObserver::Result result) {
+    if (observer_ != nullptr) {
+      observer_->OnExportFinished(result);
+    }
+  }
+
+  // Writes raw text out returning true on success. This does not escape
+  // the text in anyway.
+  bool Write(const std::string& text) {
+    if (!text.length())
+      return true;
+    size_t wrote = file_->WriteAtCurrentPos(text.c_str(), text.length());
+    bool result = (wrote == text.length());
+    if (!result) {
+      PLOG(ERROR) << "Could not write text to " << path_;
+      return false;
+    }
+    return true;
+  }
+
+  // Writes out the text string (as UTF8). The text is escaped based on
+  // type.
+  bool Write(const std::string& text, TextType type) {
+    DCHECK(base::IsStringUTF8(text));
+    std::string utf8_string;
+
+    switch (type) {
+      case ATTRIBUTE_VALUE:
+        // Convert " to &quot;
+        utf8_string = text;
+        base::ReplaceSubstringsAfterOffset(&utf8_string, 0, "\"", "&quot;");
+        break;
+
+      case CONTENT:
+        utf8_string = net::EscapeForHTML(text);
+        break;
+
+      default:
+        NOTREACHED();
+    }
+
+    return Write(utf8_string);
+  }
+
+  // Indents the current line.
+  bool WriteIndent() {
+    return Write(indent_);
+  }
+
+  // Converts a time string written to the JSON codec into a time_t string
+  // (used by bookmarks.html) and writes it.
+  bool WriteTime(const std::string& time_string) {
+    int64_t internal_value;
+    base::StringToInt64(time_string, &internal_value);
+    return Write(base::NumberToString(
+        base::Time::FromInternalValue(internal_value).ToTimeT()));
```

Comment:
`FromInternalValue` is not recommended anymore (https://source.chromium.org/chromium/chromium/src/+/master:base/time/time.h;l=769?q=FromInternalValue&ss=chromium) and may be deprecated at some point, so we should use something else for deserialization. maybe https://source.chromium.org/chromium/chromium/src/+/master:base/time/time.h;l=617?q=FromDeltaSinceWindowsEpoch&ss=chromium ?

Reply:
@diracdeltas this is code copied from upstream, I recommend we keep it as-is to avoid compatibility issues